content/blog/2020-04-dplyr-1-0-0-for-package-developers/index.Rmarkdown

```diff
@@ -0,0 +1,120 @@
+
+In this post, I want to talk a little bit about how dplyr changes might break package code, then discuss some of the major pain points a package developer might experience, and how to get help if you need it.
+
+```{r setup}
+library(dplyr, warn.conflicts = FALSE)
+```
+
+## Breaking changes
+
+There are three main ways an update to a package might break your existing code:
+
+* We've introduced a bug. Obviously, we do our best to make sure this doesn't 
+  happen (by using software development best practices like unit testing and 
+  code review) but it's impossible to eliminate all bugs.
+  
+* We've fixed a bug or otherwise made change we think is harmless. Sometimes 
+  your code accidentally depends on a behaviour that we think is incorrect and 
+  we change it. The change will be an improvement for most people, but 
+  unfortunately it [breaks your code](https://xkcd.com/1172/).
+  
+* We've deliberately made a backward incompatible interface change. We 
+  occasionally make these changes and always to significantly improve usability 
+  or consistency. Unless the package or function is experimental, we do our best 
+  to make such changes gradually, so that there's a deprecation period before 
+  the behaviour goes away altogether.
+
+dplyr 1.0.0 contains very few backward incompatible changes, but it does make a large number of changes that we believe are mostly harmless or minor improvements.  The vast majority of these will not affect data analysis code, but can affect package code, particularly unit tests. To give you a flavour for what I mean here, dplyr now preserves the names of atomic vectors:
+
+```{r}
+df <- tibble(x = c(a = 1, b = 2))
+df %>% filter(x == 1) %>% .$x %>% str()
+```
+
+(With dplyr 0.8.5, this returns an unnamed vector.)
+
+We made this change to increase internal consistency as some verbs already did preserve name, and all verbs preserved the names of list-columns. We expect that this will have minimal impact on data analysis code, but it does affect some packages because now there are names where there weren't before. If this problem affects your code, typically the best strategy is to use `unname()` to strip the names off.
+
+## Our release process
+
+To make sure all package maintainers know about potential problems, we run `R CMD check` on all 1986 packages that use dplyr. (You can see the results on [github](https://github.com/tidyverse/dplyr/tree/master/revdep#revdeps)). If your maintain a package that uses dplyr, and your package has problems, we have already emailed you several times. 
```

Comment:
```suggestion
To make sure all package maintainers know about potential problems, we run `R CMD check` on all 1,986 packages that use dplyr. (You can see the results on [github](https://github.com/tidyverse/dplyr/tree/master/revdep#revdeps)). If you maintain a package that uses dplyr, and your package has problems, we have already emailed you several times. 
```

Reply:
> we have already emailed you several times

throughout this process? I'm trying to think of a way of saying it that's not _Didn't you get my 5,000 emails?_

Reply:
> If you maintain a package that uses dplyr, and your package has problems, we have already emailed you several times throughout the release process.